### DISCUSSION TITLE
Call for agenda items: Oct. 18 Community Meeting

### DISCUSSION THREAD

**Discussion by theadactyl** [openxla/community#23] [Announcements]
Hi folks, 

What would you be most interested in discussing and/or presenting at the upcoming community meeting on Oct. 18? We are looking for the following:

- Community presentations (10-15 minutes each)
- Technical discussion topics (please be prepared to share a proposal in advance)
- Request for more information from project maintainers on XLA, StableHLO or community process. 

If you have anything in mind, please respond below OR email me at thealamkin at google dot com with your ideas by **Tuesday, Oct. 11**. 

**If you would like to add this event to your calendar, please do so by subscribing to the [OpenXLA Community Calendar](https://calendar.google.com/calendar/u/0/embed?src=c_6e3fee8576a7e330003a3130f2c89726fe487f4b3f555a12edf8a2c49005f69b@group.calendar.google.com&ctz=America/Los_Angeles) (click on the +GoogleCalendar icon in the bottom right corner).**

**Comment by ezhulenev** [comment]
I have a proof of concept for integrating Cuda Graphs + XLA runtime. Without going into too much technical details I should be able to cover the high level idea in ~15 minutes.

**Reply by nouiz** [reply]
Does it capture cublas/cudnn call?

**Reply by ezhulenev** [reply]
No. For now I only implemented outlining sequences of device kernel launches (`gpu.launch_func` from mlir gpu dialect corresponding to `mhlo.fusion` ops) into cuda graphs. But I guess cublas/cudnn should just work?

**Reply by nouiz** [reply]
Do you use graph capture or you build a cuda graph explicitly?

If you use graph capture, yes.
If you build the graph explicitly, sadly it isn't possible to build such a graph manually.
But you could do a graph capture of the library call and embed this small graph in the graph you are building manually.

So yes, it can works, but "just works" depend on how you did it.

**Reply by ezhulenev** [reply]
Graph capture for now, and rely on StreamExecutor APIs to launch kernels on stream. Just because it was the simplest option, but it should be possible to mix capture and explicit construction. 

Here is a snippet from mnist with cuda graphs:

```mlir
  func.func @a_inference_train_mnist_250__XlaMustCompil(...) {
    ...  
    call @xla.gpu.cuda.graph.launch(%c77, %c1, %c256, %arg0, %arg11, %c250, %arg3) {capture = @xla.gpu.cuda.graph.capture}
    ...
    return
  }

  func.func private @xla.gpu.cuda.graph.launch(index, index, memref<349488xi8>, index, memref<100xi8>)
    attributes {rt.custom_call = "xla.gpu.cuda.graph.launch"}
  
  rt.export @xla.gpu.cuda.graph.capture_2 ordinal 3
  func.func @xla.gpu.cuda.graph.capture_2(...) {
    %0 = arith.index_cast %arg0 : index to i32
    %1 = arith.index_cast %arg0 : index to i32
    %2 = arith.index_cast %arg0 : index to i32
    %3 = arith.index_cast %arg0 : index to i32
    %4 = arith.index_cast %arg0 : index to i32
    %5 = arith.index_cast %arg0 : index to i32
    call @xla.gpu.func.launch_0(%0, %1, %2, %3, %4, %5, %arg1, %arg2, %arg3) {kernel = "fusion_5"}
    %6 = arith.index_cast %arg4 : index to i32
    %7 = arith.index_cast %arg0 : index to i32
    %8 = arith.index_cast %arg0 : index to i32
    %9 = arith.index_cast %arg5 : index to i32
    %10 = arith.index_cast %arg0 : index to i32
    %11 = arith.index_cast %arg0 : index to i32
    call @xla.gpu.func.launch(%6, %7, %8, %9, %10, %11, %arg6, %arg7, %arg8, %arg9, %arg10, %arg11, %arg3) {kernel = "fusion_12"}
    return
  }
```

In my POC `@xla.gpu.func.launch` just launches kernel on a compute stream, but it can take a `cudaGraph` argument and do graph manipulation a bit more explicit.

**Reply by nouiz** [reply]
If you use graph capture it should just work then.
As it was in the IR and not in the GpuExecutor, I taught you where building the graph manually.

How do you manage the ptr stability issue? (Mostly, all ptr should be constant between all calls. So the input/temporary/output ptr can't change)

**Reply by ezhulenev** [reply]
My current plan is to add a cache from arguments (hash) to graph, and maybe use cudaGraphExecUpdate 
 (like in https://developer.nvidia.com/blog/employing-cuda-graphs-in-a-dynamic-environment/).

Re: capture vs construct: `@xla.gpu.cuda.graph.capture` can have explicit `graph` in the IR (passed by runtime as an argument), and instead of relying on capture it can do explicit graph construction if needed.

**Reply by sanjoy** [reply]
I believe `cudaGraphExecUpdate` won't work for cuDNN, cuBLAS etc. since it needs access to the CUDA kernels that were launched which cuDNN & cuBLAS don't directly expose.

This is going by what I remember the last I looked at CUDA graphs so would be happy to be outdated here.

**Reply by ezhulenev** [reply]
(I didn't try it myself)

Looking at example in https://developer.nvidia.com/blog/employing-cuda-graphs-in-a-dynamic-environment/ I assume that if I can capture cuDNN/BLAS calls, then I should be able to update the graph. Only the "update this graph node" APIs are not usable with cuXYZ libraries.

**Reply by nouiz** [reply]
If you "recapture and update" a captured graph, it will work.
But if you do that at every call you lose most (all?) of the speed up.

We did a cudaGraph PoC some time ago for TF/XLA.
To WAR the input ptr being dynamic issue, we captured and cached the graphs when the input ptr change.

This was working for some model when the input ptr is only a small set that always come again.
But it wasn't always working as for some models the input ptr is always a new set of ptr.
I suppose with JAX the input ptr will be more stable. But we didn't validate this.

I'll be happy to hear if you see this input ptr stability to be an issue or not with JAX.

**Reply by ezhulenev** [reply]
Yes, my hope was that pointers will not change often (didn't verify it in any real workloads yet, will not be surprised if it's not true). For this to work XLA might require more end-to-end changes and rethinking how it allocates memory.

**Reply by ezhulenev** [reply]
We don't have to capture XLA module as a whole into a graph, we can start with regions that are more likely to have stable pointers:

1. Capture only operations that have arguments inside temp allocation, presumably it should be relatively stable from run to run, or at least easier to make them stable by holding allocated temp memory
2. Capture graphs only inside loops with known trip count > N (to compensate for cache misses)

**Reply by nouiz** [reply]
That looks interesting direction.
Another potential approach: pytorch copy the model input to some constant ptr to make them stable.

If we start to hold the tmp memory ptr, we should take care to not explose the memory usage. JAX can build many small XLA module that aren't used frequently. But we can probably find a good heuristics.

Another approach would be to allocate the tmp ptr inside the graph with cudaMallocAsync. This is being optimized by the driver. The memory allocation is automatically kept unless it is needed elsewhere. This would bypass the increase of GPU memory usage issue.

**Reply by ndeep27** [reply]
@ezhulenev How can we enable the CUDA Graph Capture with XLA on a TF Saved Model?

**Comment by GleasonK** [comment]
I have started working on a StableHLO compatibility tool, to support evolving the dialect with forward and backward compatibility. I'd like to share the proposed compatibility guarantees and talk about how the tool works from a high level. Interested in community feedback / use cases. (~10min)
